### ISSUE TITLE
[BUG][Opensearch] Opensearch snapshot management: can't select AWS region for configuring s3 bucket.

### ISSUE THREAD

**Issue by hm2thr33** [opensearch-project/OpenSearch#9265]
**Describe the bug**
After upgrade Opensearch from 2.8.0 to 2.9.0 version snapshot management stop working.
Can't select AWS eu-central-1 region for configuring s3 bucket, available only one region us-east-1.

**To Reproduce**
1. Create backet in eu-central-1 region.
2. Configure backet via devtools:
```
PUT /_snapshot/backup.opensearch?pretty
{
  "type": "s3",
  "settings": {
    "bucket": "backup.opensearch",
    "region": "eu-central-1"
  }
}
```
3. Get an error:
```
"reason": "The authorization header is malformed; the region 'eu-central-1' is wrong; expecting 'us-east-1' (Service: S3, Status Code: 400, Request ID: BETMPMA86J7Q7CTH, Extended Request ID: vzSJCEIEwoZRRFkiEsKX7n/tk86XEkUnHFN28M1qxvQlrvHf7/X7NBrLSDlVpyupRoayeSfJUb8=)"
```
Expected behavior
The ability to use the bucket in the selected region, and not just in one.

Chart Name
Opensearch

**Screenshots**
N/A

**Host/Environment (please complete the following information):**
 - Helm Version: [3.12.2]
 - Kubernetes Version: [1.24.16]
 - OpenSearch Helm Chart version [2.14.1]

**Additional context**
From my side any configurations for s3 backet were changed also any configuration for opensearch.
Only upgrade Openserch version from 2.8.0 to 2.9.0.

Temporary use backet in us-east-1 region.

**Comment by raghuvanshraj** [comment]
> @raghuvanshraj care to take a look?

Taking a look

**Comment by raghuvanshraj** [comment]
@hm2thr33 after following the steps outlined by you, I am able to create a bucket in eu-central-1, take a snapshot and restore from it. Are you specifying any custom settings to the s3 client via opensearch.yml?

Pasting my requests here:
PUT snapshot repo:
```
curl --location --request PUT 'http://localhost:9200/_snapshot/bucket-9265-eu-central-1' \
--header 'Content-Type: application/json' \
--data '{
  "type": "s3",
  "settings": {
    "bucket": "bucket-9265-eu-central-1",
    "region": "eu-central-1"
  }
}'
```

PUT snapshot:
```
curl --location --request PUT 'http://localhost:9200/_snapshot/bucket-9265-eu-central-1/snapshot-2'
```

POST snapshot restore:
```
curl --location --request POST 'http://localhost:9200/_snapshot/bucket-9265-eu-central-1/snapshot-2/_restore'
```

**Comment by hm2thr33** [comment]
@raghuvanshraj I don't have any custom settings in opensearch.yml for s3 client.

The repository I use is located in the eu-central-1 region. The user has not changed. Now I'm using the same user to use a bucket in the us-east-1 region and everything works for me, but I need eu-central-1.

I also have cluster 2.8.0 not updated, I tried to connect the repository to them and everything works. But after the upgrade, I can no longer connect to the same repository, the error clearly says that I am waiting for a repository in the us-east-1 region.

**Comment by raghuvanshraj** [comment]
@hm2thr33 the requests I have pasted in the above mentioned comment are for the exact same configuration. Sometimes we see 307s for newly created buckets, but based on [this forum question](https://forum.opensearch.org/t/opensearch-repository-s3-plugin-trouble-with-aws-region-after-upgrade/15510), it looks like you are seeing a 301.

1. Can you confirm that the cluster has no overrides in the static settings?
2. Can you provide the exact response of the following payload?
```
PUT /_snapshot/backup.opensearch?pretty
{
  "type": "s3",
  "settings": {
    "bucket": "backup.opensearch",
    "region": "eu-central-1"
  }
}
```

**Comment by hm2thr33** [comment]
Hi @raghuvanshraj,

JFYI: I already have 6 clusters with this problem, all of which are running in Kubernetes, which is in the EU-CENTRAL-1 region. The buckets were also created in this region, they are not new, I started using them when I installed the first cluster on version 2.5.0, all updates to 2.9.0 were without any problems.
But on version 2.9.0, problems with the batch and region started. 

I don't have any custom settings for the S3 bucket in **opensearch.yml**:

![1](https://github.com/opensearch-project/OpenSearch/assets/37875781/43d0fcf8-da9e-4830-b52a-14eaaceaec15)

After the upgrade, when I try to add the same bucket I used for this cluster again, I see the following error:

```
{
  "error": {
    "root_cause": [
      {
        "type": "repository_verification_exception",
        "reason": "[backup.opensearch] path  is not accessible on cluster-manager node"
      }
    ],
    "type": "repository_verification_exception",
    "reason": "[backup.opensearch] path  is not accessible on cluster-manager node",
    "caused_by": {
      "type": "i_o_exception",
      "reason": "Unable to upload object [tests-tH6wHtNtTfWfntrh50S3EA/master.dat] using a single upload",
      "caused_by": {
        "type": "s3_exception",
        "reason": "The authorization header is malformed; the region 'eu-central-1' is wrong; expecting 'us-east-1' (Service: S3, Status Code: 400, Request ID: BETMPMA86J7Q7CTH, Extended Request ID: vzSJCEIEwoZRRFkiEsKX7n/tk86XEkUnHFN28M1qxvQlrvHf7/X7NBrLSDlVpyupRoayeSfJUb8=)"
      }
    }
  },
  "status": 500
}
```

I'm pretty sure I have access to this repository, because this bucket was used before the cluster upgrade, and this bucket is in the same region (EU-CENTRAL-1) as the OpenSearch cluster. The S3 bucket settings and access to them have not been changed.

**Comment by raghuvanshraj** [comment]
I was able to replicate this issue, but for me it is only happening if the bucket region and the region supplied in the request are different. So if I create a bucket `bucket-9265-ap-northeast-1` in `ap-northeast-1` and then try to register a repository against it in the region `eu-central-1`
```
curl --location --request PUT 'http://localhost:9200/_snapshot/bucket-9265-eu-central-1' \
--header 'Content-Type: application/json' \
--data '{
  "type": "s3",
  "settings": {
    "bucket": "bucket-9265-ap-northeast-1",
    "region": "eu-central-1"
  }
}'
```

I get the following error:
```
{
    "error": {
        "root_cause": [
            {
                "type": "repository_verification_exception",
                "reason": "[bucket-9265-eu-central-1] path  is not accessible on cluster-manager node"
            }
        ],
        "type": "repository_verification_exception",
        "reason": "[bucket-9265-eu-central-1] path  is not accessible on cluster-manager node",
        "caused_by": {
            "type": "i_o_exception",
            "reason": "Unable to upload object [tests-LdGyFSq_RNeY3eWEW467Xg/master.dat] using a single upload",
            "caused_by": {
                "type": "s3_exception",
                "reason": "The authorization header is malformed; the region 'eu-central-1' is wrong; expecting 'ap-northeast-1' (Service: S3, Status Code: 400, Request ID: B8E5HHZFM04XG7FF, Extended Request ID: zqtCpaxyP19GvIGgTDOZiUHgpDMavGAhf/qiTFiadBP2oEVlAkJRUIevjI5HBK1EXZ0ulyal2rg=)"
            }
        }
    },
    "status": 500
}
```

Similarly, if I don't supply a region in my payload, it defaults to `us-east-1` since we don't provide a region override to the client:
```
curl --location --request PUT 'http://localhost:9200/_snapshot/bucket-9265-eu-central-1' \
--header 'Content-Type: application/json' \
--data '{
  "type": "s3",
  "settings": {
    "bucket": "bucket-9265-ap-northeast-1"
  }
}'
```

As expected, I get the same error, but instead of `eu-central-1`, I get `us-east-1`:
```
{
    "error": {
        "root_cause": [
            {
                "type": "repository_verification_exception",
                "reason": "[bucket-9265-eu-central-1] path  is not accessible on cluster-manager node"
            }
        ],
        "type": "repository_verification_exception",
        "reason": "[bucket-9265-eu-central-1] path  is not accessible on cluster-manager node",
        "caused_by": {
            "type": "i_o_exception",
            "reason": "Unable to upload object [tests-3W0j4lyLRte8NdwHALAnaQ/master.dat] using a single upload",
            "caused_by": {
                "type": "s3_exception",
                "reason": "The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'ap-northeast-1' (Service: S3, Status Code: 400, Request ID: ET97XVJ1C3PVEPX9, Extended Request ID: Zx0SKdusKV1YWvjHv7rLsCt3M/NIyINTNPaJi5WUtdqPRQkKRxsUIXxjOj/zOekdSIaoinFIisk=)"
            }
        }
    },
    "status": 500
}
```

Based on this, the only possible scenario I can see here is that S3 is resolving the bucket `backup.opensearch` to be in `us-east-1`, not `eu-central-1`, due to which when a client scoped to `eu-central-1` calls S3 service hoping to upload objects to it, S3 throws this error.

**Comment by dblock** [comment]
Thanks for debugging this @raghuvanshraj - @hm2thr33 care to take a look at the details above?

**Comment by sachinpkale** [comment]
@hm2thr33 Closing this issue. Feel free to re-open if problem still exists.

**Comment by rtsui-mk** [comment]
Hi @sachinpkale , This problem is not solved in Opensearch 2.19

**Comment by hm2thr33** [comment]
@ICHx try add `s3.client.default.region` and `s3.client.default.endpoint` at your `opensearch.yml`

like this
`s3.client.default.region: eu-central-1`
`s3.client.default.endpoint: https://s3.eu-central-1.amazonaws.com/`

it worked for me

**Comment by rtsui-mk** [comment]
Hi @hm2thr33 ,

Thanks, I know the flag as I checked above too.

This is undesirable as it requires restarting the cluster and doesn't support multiple repo multiple region.

I added mention in the FR I found

Thanks

**Comment by rtsui-mk** [comment]
I see, I checked your thread in the forum

While I might not need cross region access, it would be highly favourable if it can be dynamically altered.

```
2025-07-18 02:51:52,680 WARNING: PUT http://localhost:9200/_cluster/settings [status:400 request:0.004s]
2025-07-18 02:51:52,682 ERROR: RequestError(400, 'illegal_argument_exception', 'persistent setting [s3.client.default.region], not dynamically updateable')
```


> In https://github.com/opensearch-project/OpenSearch/pull/7372 we have migrated discovery-ec2 and repository-s3 plugins to AWS SDK v2. Until then, AWS SDK v1 allowed for cross-region access, and AWS SDK v2 no longer does ([ref](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/migration-whats-different.html)). We need to specify the correct region for the S3 bucket used in integration tests.

**Comment by rtsui-mk** [comment]
From here I am seeing cross-region access is actually possible.  I am thinking if this ticket can be re-opened.

https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/s3-cross-region.html

**Comment by rtsui-mk** [comment]
Hi @hm2thr33,

I tried that I can specify `endpoint` and `region` in es repo settings and then it worked like it used to.  No need editing opensearch.yml in that way.  And in this way no need turning on cross-region access although it is desirable.

This was undocumented in Opensearch's doc but mentioned in Elasticsearch's.

```
              es_settings['endpoint'] = f'https://s3.{ es_settings.get("region", "us-west-2") }.amazonaws.com'
              es_repo_params = {'type': 's3', 'settings': es_settings}

              logger.info('S3: creating ES repo %s with %s',repo, es_repo_params)

              status:dict ={}
              try:
                api_resp = self.ES_API.snapshot.create_repository(repository=repo,
                                                                  body=es_repo_params)
```
Here is my implementation involved.  Without the endpoint, error is observed.

---
(Unrelated to the fix)
I also patched the plugin to allow cross region access (available since 2.20.111)The error message becomes  

> Please re-send this request to the specified temporary endpoint. Continue to use the original request endpoint for future requests.

While when using awscli --debug we can see the same message, opensearch s3 client is unable to handle redirection.

```
vagrant@07c09107d183 tests → curl 0:19200/_snapshot/test-repo -H 'Content-Type: application/json' -d '{"type": "s3", "settings": {"bucket": "vos-elk-local-build-h55jk7fq96osa0jk"}}' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   843  100   765  100    78   2602    265 --:--:-- --:--:-- --:--:--  2867
{
  "error": {
    "root_cause": [
      {
        "type": "repository_verification_exception",
        "reason": "[test-repo] path  is not accessible on cluster-manager node"
      }
    ],
    "type": "repository_verification_exception",
    "reason": "[test-repo] path  is not accessible on cluster-manager node",
    "caused_by": {
      "type": "i_o_exception",
      "reason": "Unable to upload object [tests-N0zLqlXVS8msemkaoG8eEQ/master.dat] using a single upload",
      "caused_by": {
        "type": "s3_exception",
        "reason": "Please re-send this request to the specified temporary endpoint. Continue to use the original request endpoint for future requests. (Service: S3, Status Code: 307, Request ID: 4Q7KKXR1M6F9M5AB, Extended Request ID: Y6dL/VC6wLPTOY4aEt6oG5/vEQotWxgS6fwvMmLk9WQZmcBPNt1Hiu9SiztIqE8pzrx2+MrcEuYprPdTLTPQGLCAwN0DGNDh)"
      }
    }
  },
  "status": 500
}
```


I searched and retried in awscli,
```
❯ aws s3 ls --region us-east-1 [s3://bucket-sndn51jcaxn3ot6j/](s3://bucket-sndn51jcaxn3ot6j/) --debug

2025-07-22 17:28:40,208 - MainThread - botocore.parsers - DEBUG - Response body:
b'<?xml version="1.0" encoding="UTF-8"?>\n<Error><Code>PermanentRedirect</Code><Message>The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.</Message><Endpoint>bucket-sndn51jcaxn3ot6j.s3-us-west-1.amazonaws.com</Endpoint><Bucket>bucket-sndn51jcaxn3ot6j</Bucket><RequestId>BQN93AHFQJHZRTZY</RequestId><HostId>iNduI12CIQOjWg4JFBVvk6GlKFfe7PDJa83qT/4zeGy/Hfr90AbgeEXWLDJregT6MRd2Sjrxsrvgkze+RQlyBhEOL45kcLKcbAOX11KUnek=</HostId></Error>'
```

In awscli it's just a debug message and was able to redirect itself, but not opensearch

The bucket is in us-west-1

`aws s3 ls --region us-west-1 [s3://bucket-sndn51jcaxn3ot6j/](s3://bucket-sndn51jcaxn3ot6j/) --debug`

Using the same region in cli do not have that message
